### DISCUSSION TITLE
BE[201] Node.js 連接 Mysql 出現錯誤"Error: connect ECONNREFUSED 127.0.0.1:3306"

### DISCUSSION THREAD

**Discussion by rich2020s** [Lidemy/forum#168] [後端相關問題] [answered]
## 問題：
用 Node.js 連線到 Mysql 出現錯誤。
作業系統為 macOS BigSur。
### 我的程式碼：
![image](https://raw.githubusercontent.com/rich2020s/pic/main/errMessage.png)

### 錯誤訊息：
![image](https://raw.githubusercontent.com/rich2020s/pic/main/connectcode.png)

### 目前掌握的裝況：
* 密碼沒錯
* 用 php 可以正常連線到 Mysql
*  3306 port 有開

### 找到的資料
https://github.com/mysqljs/mysql/issues/718
https://github.com/mysqljs/mysql/issues/2268
[解决Node.js使用MySQL出现connect ECONNREFUSED 127.0.0.1:3306的问题](https://blog.csdn.net/Zhooson/article/details/107337987)
[econnrefused MySQL_解决Node.js使用MySQL出现connect ECONNREFUSED 127.0.0.1:3306的问题](https://blog.csdn.net/weixin_33548183/article/details/113416989?utm_medium=distribute.pc_relevant.none-task-blog-2~default~baidujs_title~default-0.pc_relevant_baidujshouduan&spm=1001.2101.3001.4242)
有在 mysql 的issues 上看到很多類似的問題，但大部份方法都試過了無效。
有一個沒試過，是加上 `socketPath`:[`mysqld.sock`的路徑]。但我不知道要怎麼把路徑設在 xampp 打開的空間裡面。直接輸入路徑 `/opt/lampp/var/mysql/mysql.sock` 的話會找不到。
然後找到一篇狀況類似的 [Error connect ENOENT /var/run/mysqld/mysqld.sock in node.js](https://stackoverflow.com/questions/40455815/error-connect-enoent-var-run-mysqld-mysqld-sock-in-node-js)
但看不太懂改了什麼。我的`my.cnf` 沒有`bind-address` 這行。host 的 ip 試了 xampp 的 ip，換成出別的錯誤訊息。
也不確定這個方向有沒有正確。求指點～
謝謝看完的各位！

**Comment by aszx87410** [comment, marked as answer]
我想起來了，這好像上一期的常見問題之一！因為你 PHP 跟 Mysql 都是跑在虛擬機裡面，所以用 127.0.0.1 連得到，但我們現在 JS 是跑在本機，mysql 是跑在虛擬機，所以其實是不同地方

解法有兩種，第一種是 host 改填虛擬機 ip 看看，第二種是 network 那邊增加一筆 3306 port 的對應，把虛擬機的 3306 port 導到本機的 3306

**Reply by rich2020s** [reply]
感謝老師，後來用第一種方法成功解決了。
另外感謝 @rockyooooooo 提醒，這個問題其實在第十七週的[課程簡介](https://lidemy.com/courses/1360789/lectures/31285005)就有講要怎麼設定了。有相同問題的同學可以直接參考。
謝謝大家～
